newlib/libc/sys/vita/select.c

```diff
@@ -0,0 +1,138 @@
+/*
+
+Copyright (C) 2016, Sunguk Lee
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+
+*/
+#include <sys/select.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <psp2/kernel/threadmgr.h>
+#include <psp2/net/net.h>
+#include <psp2/types.h>
+
+#include "vitadescriptor.h"
+
+#define SCE_ERRNO_MASK 0xFF
+
+#define MAX_EVENTS 255
+
+int select(int nfds, fd_set *readfds, fd_set *writefds, fd_set *exceptfds,
+		   struct timeval *timeout)
+{
+	uint64_t wait = timeout->tv_sec * 1000000 + timeout->tv_usec;
+
+	if (nfds == 0) {
+		sceKernelDelayThread(wait);
+		return 0;
+	}
+
+	fd_set lock = {0};
+
+	int i;
+	int res;
+	int eid = sceNetEpollCreate("", 0);
+	if (eid < 0) {
+		res = (int)eid;
+		goto exit;
+	}
+	for (i = 0; i < nfds; i++) {
+		SceNetEpollEvent ev = {0};
+		ev.data.fd = i;
+		if (readfds && FD_ISSET(i, readfds)) {
+			ev.events |= SCE_NET_EPOLLIN;
+		}
+		if (writefds && FD_ISSET(i, writefds)) {
+			ev.events |= SCE_NET_EPOLLOUT;
+		}
+		if (exceptfds && FD_ISSET(i, exceptfds)) {
+			ev.events |= SCE_NET_EPOLLERR;
+		}
+		if (ev.events == 0) {
+			continue;
+		}
+
+		DescriptorTranslation *fdmap = __vita_fd_grab(i);
```

Comment:
It can race condition between this point and the comment left here: https://github.com/vitasdk/newlib/pull/9/files#r80349878

The problem lies under the use of ``__vita_fdmap``, which may no longer point to the correct fd. You must drop the pointer returned from ``__vita_fd_grab``.